### PR TITLE
4164 make message routing configurable

### PR DESCRIPTION
**Pull request by weskubo-cgi**
The goal of this Story was to be able to switch how messages were routed without having to make code changes. The approach is described here: https://proactionca.ent.cgi.com/confluence/display/BCMOHAD/HNSecure+-+Routing+Configuration.

As part of this change the main Route was broken down into sub Routes which are conditionally invoked based on the message type (or grouping thereof). This was necessary as the main Route would have become unwieldly with the additional code changes.
Unfortunately, there were two side effects of this:
- Exception handling is only applied to Routes in the current RouteBuilder (i.e. class). In order to replicate the exception handling across all Routes we made use of an abstract base class
- The onCompetion() handler was being invoked for each complete Route, which could be up to 3 Routes for something like HIB. The onCompletion() logic was moved to a ResponseHandlerRoute that gets invoked at the end of the main Route as well as on any Exception scenarios.

The latter change resulted in some odd behaviour in the unit test whereby the JSON formatting was slightly different (though the content was equivalent). I verified that the "new" response actually matches what is returned in the live application and updated the expected result accordingly.

Lastly, Bug 4377 was incorporated into this fix as I had to test exception handling.

I also had to incorporate the changes from 4367 into this branch so please keep an eye out for merge issues.

### CODE REVIEW

File: hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/json/pharmanet/ProcessV2ToPharmaNetJson.java [hunk 1]
```diff
@@ -10,7 +10,6 @@
 
 import ca.bc.gov.hlth.hncommon.util.LoggingUtil;
 import ca.bc.gov.hlth.hnsecure.exception.CustomHNSException;
-import ca.bc.gov.hlth.hnsecure.parsing.Util;
```
Comment (weskubo-cgi) [inline hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/json/pharmanet/ProcessV2ToPharmaNetJson.java:13]: Unrelated code cleanup.

File: hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/messagevalidation/ExceptionHandler.java [hunk 2]
```diff
@@ -59,7 +61,7 @@ public void process(Exchange exchange) {
 				handleException(exchange, ErrorMessage.HL7Error_Msg_Unknown, HttpStatus.BAD_REQUEST_400, TransactionEventType.ERROR);
 				break;
 			}
-		} else if (exception instanceof HttpHostConnectException) {
+		} else if (exception instanceof HttpHostConnectException || exception instanceof UnknownHostException) {
```
Comment (weskubo-cgi) [inline hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/messagevalidation/ExceptionHandler.java:64]: This is the fix for Bug 4377

File: hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/parsing/FormatRTransResponse.java [hunk 3]
```diff
@@ -7,7 +7,6 @@
 import org.slf4j.LoggerFactory;
 
 import ca.bc.gov.hlth.hncommon.util.LoggingUtil;
-import ca.bc.gov.hlth.hnsecure.exception.CustomHNSException;
```
Comment (weskubo-cgi) [inline hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/parsing/FormatRTransResponse.java:10]: Unrelated code cleanup.

File: hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/parsing/ProtocolEvaluator.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+	private static final ApplicationProperties properties = ApplicationProperties.getInstance();
+	
+	@Override
+	public void process(Exchange exchange) throws Exception {
+		final String methodName = LoggingUtil.getMethodName();
+		
+		String messageType = (String)exchange.getProperty(Util.PROPERTY_MESSAGE_TYPE);
+		
+		// Dynamically look up the protocol from the properties
+		String protocol = properties.getValue(messageType + PROTOCOL_PREFIX);
```
Comment (weskubo-cgi) [inline hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/parsing/ProtocolEvaluator.java:28]: A null/empty value here will be handled in the Route.

File: hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/routes/HandleResponseRoute.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import ca.bc.gov.hlth.hnsecure.parsing.Util;
+
+public class HandleResponseRoute extends RouteBuilder {
+
+	@Override
+	public void configure() throws Exception {
+		from("direct:handleResponse").routeId("handle-response-route")
+			// create filedrops if enabled
+	    	.choice().when(exchangeProperty(Util.PROPERTY_IS_FILE_DROPS_ENABLED).isEqualToIgnoreCase(Boolean.TRUE))
+	    		.bean(ResponseFileDropGenerater.class).id("ResponseFileDropGenerater")
```
Comment (tschia) [inline hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/routes/HandleResponseRoute.java:20]: it looks like here we call the filedrop directly, and elsewhere we wiretap it to "direct:start" (poorly named route)
Can we use the same approach for both?

So something like 
to("direct:requestFiledrop")
to("direct:responseFiledrop")

this could maybe be a later bug fix/refactor

Comment (weskubo-cgi) [reply hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/routes/HandleResponseRoute.java:20]: I'm having a look. If it's simple (no side effects) I'll incorporate it here.

File: hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/routes/HIBCRoute.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+			.choice()
+				.when(exchangeProperty(Util.PROPERTY_MESSAGE_PROTOCOL).isEqualTo(Util.PROTOCOL_HTTP))
+					.to("direct:hibcHTTP")
+				.when(exchangeProperty(Util.PROPERTY_MESSAGE_PROTOCOL).isEqualTo(Util.PROTOCOL_MQ))
+					.to("direct:hibcMQ")
+				.otherwise()
+					.log("Protocol for HIBC message type ${exchangeProperty.messageType} not found or not valid. Defaulting to MQ")
+					.to("direct:hibcMQ")
+			.end();
+		// XXX This is currently just a simple HTTP route which will need additional configuration (auth, request/response conversion)
```
Comment (daveb-hni) [inline hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/routes/HIBCRoute.java:43]: Add a line for format consistency.

**Review by daveb-hni — COMMENTED** [review]
Minor format comments.

File: hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/routes/HandleResponseRoute.java (hunk 5 above)
Comment (weskubo-cgi) [reply hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/routes/HandleResponseRoute.java]: This was completed. It was straightforward.

File: hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/parsing/ProtocolEvaluator.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+import com.nimbusds.oauth2.sdk.util.StringUtils;
+
+import ca.bc.gov.hlth.hncommon.util.LoggingUtil;
+import ca.bc.gov.hlth.hnsecure.properties.ApplicationProperties;
+
+public class ProtocolEvaluator implements Processor {
+
+	private static final Logger logger = LoggerFactory.getLogger(ProtocolEvaluator.class);
+	
+	private static final String PROTOCOL_PREFIX = ".protocol";
```
[the remark this answers is not in the document]
Comment (weskubo-cgi) [reply hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/parsing/ProtocolEvaluator.java]: Renamed to PROTOCOL_SUFFIX.

File: hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/routes/HIBCRoute.java (hunk 6 above)
Comment (weskubo-cgi) [reply hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/routes/HIBCRoute.java:44]: Spacing added.

**Review by daveb-hni — APPROVED** [review]
Looks ready for merge
